yt_dlp/extractor/tumblr.py

```diff
@@ -256,23 +387,41 @@ def _perform_login(self, username, password):
         if not self._ACCESS_TOKEN:
             return
 
-        self._download_json(
-            self._OAUTH_URL, None, 'Logging in',
-            data=urlencode_postdata({
-                'password': password,
-                'grant_type': 'password',
-                'username': username,
-            }), headers={
-                'Content-Type': 'application/x-www-form-urlencoded',
-                'Authorization': f'Bearer {self._ACCESS_TOKEN}',
-            },
-            errnote='Login failed', fatal=False)
+        data = {
+            'password': password,
+            'grant_type': 'password',
+            'username': username,
+        }
+        if self.get_param('twofactor'):
+            data['tfa_token'] = self.get_param('twofactor')
+
+        def _call_login():
+            return self._download_json(
+                self._OAUTH_URL, None, 'Logging in',
+                data=urlencode_postdata(data),
+                headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                    'Authorization': f'Bearer {self._ACCESS_TOKEN}',
+                },
+                errnote='Login failed', fatal=False,
+                expected_status=lambda s: s == 200 or 400 <= s < 500)
+
+        response = _call_login()
+        if response.get('error') == 'tfa_required':
```

Comment:
Could be better to make error handling non-fatal itself:
```suggestion
        if traverse_obj(response, 'error') == 'tfa_required':
```